Date: Wed, 7 Aug 2024 16:40:11 +0200
Subject: scsi: mpi3mr: Avoid memcpy field-spanning write WARNING

[ Upstream commit 429846b4b6ce9853e0d803a2357bb2e55083adf0 ]

When the "storcli2 show" command is executed for eHBA-9600, mpi3mr driver
prints this WARNING message:

  memcpy: detected field-spanning write (size 128) of single field "bsg_reply_buf->reply_buf" at drivers/scsi/mpi3mr/mpi3mr_app.c:1658 (size 1)
  WARNING: CPU: 0 PID: 12760 at drivers/scsi/mpi3mr/mpi3mr_app.c:1658 mpi3mr_bsg_request+0x6b12/0x7f10 [mpi3mr]

The cause of the WARN is 128 bytes memcpy to the 1 byte size array "__u8
replay_buf[1]" in the struct mpi3mr_bsg_in_reply_buf. The array is intended
to be a flexible length array, so the WARN is a false positive.

To suppress the WARN, remove the constant number '1' from the array
declaration and clarify that it has flexible length. Also, adjust the
memory allocation size to match the change.

Suggested-by: Sathya Prakash Veerichetty <sathya.prakash@broadcom.com>
Signed-off-by: Shin'ichiro Kawasaki <shinichiro.kawasaki@wdc.com>
Link: https://lore.kernel.org/r/20240323084155.166835-1-shinichiro.kawasaki@wdc.com
Signed-off-by: Martin K. Petersen <martin.petersen@oracle.com>
[patch it against out-of-source driver, struct mpi3mr_bsg_in_reply_buf
is in mpi3mr_app.h instead of scsi_bsg_mpi3mr.h]
Signed-off-by: Mahmoud Adam <mngyadam@amazon.com>

diff --git a/drivers/amazon/scsi/mpi3mr/mpi3mr_app.c b/drivers/amazon/scsi/mpi3mr/mpi3mr_app.c
--- a/drivers/amazon/scsi/mpi3mr/mpi3mr_app.c
+++ b/drivers/amazon/scsi/mpi3mr/mpi3mr_app.c
@@ -2803,7 +2803,7 @@ static long mpi3mr_bsg_process_mpt_cmds(struct bsg_job *job)
 	if ((mpirep_offset != 0xFF) &&
 	    drv_bufs[mpirep_offset].bsg_buf_len) {
 		drv_buf_iter = &drv_bufs[mpirep_offset];
-		drv_buf_iter->kern_buf_len = (sizeof(*bsg_reply_buf) - 1 +
+		drv_buf_iter->kern_buf_len = (sizeof(*bsg_reply_buf) +
 					   mrioc->reply_sz);
 		bsg_reply_buf = kzalloc(drv_buf_iter->kern_buf_len, GFP_KERNEL);
 
diff --git a/drivers/amazon/scsi/mpi3mr/mpi3mr_app.h b/drivers/amazon/scsi/mpi3mr/mpi3mr_app.h
--- a/drivers/amazon/scsi/mpi3mr/mpi3mr_app.h
+++ b/drivers/amazon/scsi/mpi3mr/mpi3mr_app.h
@@ -375,7 +375,7 @@ struct mpi3mr_bsg_in_reply_buf {
 	uint8_t mpi_reply_type;
 	uint8_t rsvd1;
 	uint16_t rsvd2;
-	uint8_t reply_buf[1]; /*Variable Length buffer based on mpi reply type*/
+	uint8_t reply_buf[]; /*Variable Length buffer based on mpi reply type*/
 };
 /**
  * struct mpi3mr_buf_entry - User buffer descriptor for MPI
